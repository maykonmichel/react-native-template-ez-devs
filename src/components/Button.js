import React, { useMemo } from 'react';
import { ActivityIndicator, Platform, StyleSheet, ViewPropTypes } from 'react-native';
import PropTypes from 'prop-types';
import { toUpper } from 'lodash';

import colors from '../configs/colors';
import Icon from './Icon';
import Text from './Text';
import TouchableView from './TouchableView';

const Button = ({ disabled, icon, loading, style, title, titleAttributes, ...attributes }) => {
  const normalizedTitle = useMemo(() => (Platform.OS === 'android' ? toUpper(title) : title), []);

  return (
    <TouchableView
      activeOpacity={0.8}
      disabled={disabled}
      style={StyleSheet.flatten([styles.container, style, disabled && styles.disabledContainer])}
      {...attributes}
    >
      <If condition={loading}>
        <ActivityIndicator animating color="white" />
      </If>
      <If condition={icon}>
        <Icon name={icon} size={20} color={disabled ? colors.light : '#fff'} />
      </If>
      <Text
        style={StyleSheet.flatten([styles.title, disabled && styles.disabledTitle])}
        {...titleAttributes}
      >
        {normalizedTitle}
      </Text>
    </TouchableView>
  );
};

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: colors.primary,
    marginTop: 10,
    marginBottom: 30,
    paddingVertical: 8,
    paddingHorizontal: 30,
    shadowColor: '#000',
    shadowOffset: {
      width: 0,
      height: 5
    },
    shadowOpacity: 0.34,
    shadowRadius: 6.27,

    elevation: 10
  },
  disabledContainer: {
    backgroundColor: colors.secondary
  },
  title: {
    color: '#fff',
    textAlign: 'center'
  },
  disabledTitle: {
    color: colors.shine
  }
});

Button.propTypes = {
  disabled: PropTypes.bool,
  icon: PropTypes.string,
  loading: PropTypes.bool,
  style: ViewPropTypes.style,
  title: PropTypes.string,
  titleAttributes: PropTypes.shape(Text.propTypes)
};

Button.defaultProps = {
  disabled: false,
  icon: undefined,
  loading: false,
  style: {},
  title: '',
  titleAttributes: {}
};

export default React.memo(Button);
