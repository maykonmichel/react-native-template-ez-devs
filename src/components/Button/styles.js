import { StyleSheet } from 'react-native';

import colors from '../../configs/colors';

export default StyleSheet.create({
  button: {
    flexDirection: 'row',
    width: '80%',
    justifyContent: 'center',
    alignItems: 'center',
    padding: 8,
    borderRadius: 3,
    backgroundColor: colors.primary
  },
  disabled: {
    backgroundColor: colors.secondary
  },
  disabledTitle: {
    color: colors.light
  },
  light: {
    backgroundColor: 'transparent',
    elevation: 0
  },
  loading: {
    marginRight: 15
  },
  title: {
    textAlign: 'center',
    paddingTop: 2,
    paddingBottom: 1
  }
});
