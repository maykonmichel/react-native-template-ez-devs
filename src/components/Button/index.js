import React from 'react';
import {
  ActivityIndicator,
  Platform,
  StyleSheet,
  TouchableNativeFeedback,
  TouchableOpacity,
  View,
  ViewPropTypes
} from 'react-native';
import PropTypes from 'prop-types';

import styles from './styles';
import Text from '../Text';

const Touchable = Platform.OS === 'android' ? TouchableNativeFeedback : TouchableOpacity;

export default class Button extends React.PureComponent {
  static propTypes = {
    disabled: PropTypes.bool,
    disabledStyle: ViewPropTypes.style,
    disabledTitleStyle: PropTypes.shape(Text.propTypes),
    leftComponent: PropTypes.element,
    light: PropTypes.bool,
    loading: PropTypes.bool,
    loadingAttributes: PropTypes.shape(ActivityIndicator.propTypes),
    rightComponent: PropTypes.element,
    style: ViewPropTypes.style,
    title: PropTypes.string,
    titleStyle: PropTypes.shape(Text.propTypes)
  };

  static defaultProps = {
    disabled: false,
    disabledStyle: {},
    disabledTitleStyle: {},
    leftComponent: null,
    light: false,
    loading: false,
    loadingAttributes: {},
    rightComponent: null,
    style: {},
    title: '',
    titleStyle: {}
  };

  render() {
    const {
      disabled,
      disabledStyle,
      disabledTitleStyle,
      leftComponent,
      light,
      loading,
      loadingAttributes,
      loadingStyles,
      rightComponent,
      style,
      title,
      titleStyle,
      ...attributes
    } = this.props;
    return (
      <Touchable
        activeOpacity={light ? 0.4 : 0.8}
        disabled={disabled}
        underlayColor={light ? 'transparent' : undefined}
        {...attributes}
      >
        <View
          style={StyleSheet.flatten([
            styles.button,
            style,
            light && styles.light,
            disabled && styles.disabled,
            disabled && disabledStyle
          ])}
        >
          <If condition={loading}>
            <ActivityIndicator
              style={StyleSheet.flatten([styles.loading, loadingStyles])}
              animating
              color="white"
              {...loadingAttributes}
            />
          </If>
          {leftComponent}
          <Text
            color="shine"
            style={StyleSheet.flatten([
              styles.title,
              titleStyle,
              disabled && styles.disabledTitle,
              disabled && disabledTitleStyle
            ])}
          >
            {title}
          </Text>
          {rightComponent}
        </View>
      </Touchable>
    );
  }
}
