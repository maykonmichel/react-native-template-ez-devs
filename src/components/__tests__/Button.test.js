import React from 'react';
import { shallow } from 'enzyme';

import Button from '../Button';

describe('Testing Button component', () => {
  const button = shallow(<Button>Test</Button>);
  it('renders as expected', () => {
    expect(button).toMatchSnapshot();
  });
  it('calls function on press', () => {
    const onPress = jest.fn();
    button.setProps({ onPress });
    button.simulate('press');
    expect(onPress.mock.calls.length).toBe(1);
  });
  it('renders with loading indicator', () => {
    button.setProps({ loading: true });
    expect(button).toMatchSnapshot();
  });
  it('renders disabled', () => {
    button.setProps({ disabled: true });
    expect(button).toMatchSnapshot();
  });
});
