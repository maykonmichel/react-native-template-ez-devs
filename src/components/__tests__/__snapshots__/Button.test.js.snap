// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Button component renders as expected 1`] = `
<Memo(TouchableView)
  activeOpacity={0.8}
  disabled={false}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#141E50",
      "elevation": 10,
      "flexDirection": "row",
      "justifyContent": "center",
      "marginBottom": 30,
      "marginTop": 10,
      "paddingHorizontal": 30,
      "paddingVertical": 8,
      "shadowColor": "#000",
      "shadowOffset": Object {
        "height": 5,
        "width": 0,
      },
      "shadowOpacity": 0.34,
      "shadowRadius": 6.27,
    }
  }
>
  <Memo(Text)
    style={
      Object {
        "color": "#fff",
        "textAlign": "center",
      }
    }
  />
</Memo(TouchableView)>
`;

exports[`Testing Button component renders disabled 1`] = `
<Memo(TouchableView)
  activeOpacity={0.8}
  disabled={true}
  onPress={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#6c757d",
      "elevation": 10,
      "flexDirection": "row",
      "justifyContent": "center",
      "marginBottom": 30,
      "marginTop": 10,
      "paddingHorizontal": 30,
      "paddingVertical": 8,
      "shadowColor": "#000",
      "shadowOffset": Object {
        "height": 5,
        "width": 0,
      },
      "shadowOpacity": 0.34,
      "shadowRadius": 6.27,
    }
  }
>
  <ActivityIndicator
    animating={true}
    color="white"
    hidesWhenStopped={true}
    size="small"
  />
  <Memo(Text)
    style={
      Object {
        "color": "#f1f8ff",
        "textAlign": "center",
      }
    }
  />
</Memo(TouchableView)>
`;

exports[`Testing Button component renders with loading indicator 1`] = `
<Memo(TouchableView)
  activeOpacity={0.8}
  disabled={false}
  onPress={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#141E50",
      "elevation": 10,
      "flexDirection": "row",
      "justifyContent": "center",
      "marginBottom": 30,
      "marginTop": 10,
      "paddingHorizontal": 30,
      "paddingVertical": 8,
      "shadowColor": "#000",
      "shadowOffset": Object {
        "height": 5,
        "width": 0,
      },
      "shadowOpacity": 0.34,
      "shadowRadius": 6.27,
    }
  }
>
  <ActivityIndicator
    animating={true}
    color="white"
    hidesWhenStopped={true}
    size="small"
  />
  <Memo(Text)
    style={
      Object {
        "color": "#fff",
        "textAlign": "center",
      }
    }
  />
</Memo(TouchableView)>
`;
